### PR TITLE
Buffs the Prince of Terror

### PR DESCRIPTION
**Pull request by DGamerL**
## What Does This PR Do
Makes the prince have no slowdown - as fast as humans
Has him spawn in a vent instead. HE DOES NOT GET VENTCRAWL

## Why It's Good For The Game
Prince has for long been a meh major midround, being defeated if the crew is competent and stays out of range, and just pelts it with ranged weapons.
This should bring it up to speed (Hah) and spawning in a vent gives it the freedom to choose where it spawns.

## Testing
Prince of terror go vroom vroom

## Changelog
:cl:
tweak: Prince of Terror did leg day eight days per week, and is now faster
tweak: Prince of Terror now spawns in a vent
/:cl:

### CODE REVIEW

**Comment by Miraviel** [comment]
Might want to consider giving him a spit with a huge cooldown as well if you seek to buff it, so he can get unstuck when grav goes out. Today we had a round where a prince appeared, someone turned off the gravgen, prince got stuck and just got gunned down while he was surprisedpikachu staring at people. (Blobbernauts have the same issue but that's a different issue)

**Comment by DGamerL** [comment]
> Might want to consider giving him a spit with a huge cooldown as well if you seek to buff it, so he can get unstuck when grav goes out. Today we had a round where a prince appeared, someone turned off the gravgen, prince got stuck and just got gunned down while he was surprisedpikachu staring at people. (Blobbernauts have the same issue but that's a different issue)

Based on some feedback (That I forgot to put in the description, oops) the plan is that I'm gonna give the Prince a charge ability, like gargs have, which should also fix that same issue :)

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm [hunk 1]
```diff
@@ -40,3 +42,34 @@
...
+	. = ..()
+	if(mind)
+		var/obj/effect/proc_holder/spell/spell = new /obj/effect/proc_holder/spell/princely_charge(null)
+		mind.AddSpell(spell)
+	else
+		RegisterSignal(src, COMSIG_MOB_LOGIN, TYPE_PROC_REF(/mob/living/simple_animal/hostile/poison/terror_spider/prince, give_spell))
+
+/mob/living/simple_animal/hostile/poison/terror_spider/prince/proc/give_spell()
+	SIGNAL_HANDLER
+	var/obj/effect/proc_holder/spell/spell = new /obj/effect/proc_holder/spell/princely_charge(null)
```
Comment (lewcc) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:56]:
```suggestion
	var/obj/effect/proc_holder/spell/spell = new /obj/effect/proc_holder/spell/princely_charge()
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm [hunk 2]
```diff
@@ -40,3 +42,34 @@
 	L.KnockDown(10 SECONDS)
 	L.adjustStaminaLoss(40)
 	return ..()
+
+/mob/living/simple_animal/hostile/poison/terror_spider/prince/Initialize(mapload)
+	. = ..()
+	if(mind)
+		var/obj/effect/proc_holder/spell/spell = new /obj/effect/proc_holder/spell/princely_charge(null)
```
Comment (lewcc) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:49]: This should be equivalent to passing nothing in
```suggestion
		var/obj/effect/proc_holder/spell/spell = new /obj/effect/proc_holder/spell/princely_charge()
```

File: code/modules/mob/living/living.dm [hunk 3]
```diff
@@ -1167,3 +1167,36 @@
 	C.take_organ_damage(damage)
 	C.KnockDown(3 SECONDS)
 	C.visible_message("<span class='danger'>[C] crashes into [src], knocking them both over!</span>", "<span class='userdanger'>You violently crash into [src]!</span>")
+
+/mob/living/throw_impact(atom/hit_atom, throwingdatum)
+	. = ..()
+	if(has_status_effect(STATUS_EFFECT_CHARGING))
```
Comment (lewcc) [inline code/modules/mob/living/living.dm:1173]: You can probably early return here if it has either of the two status effects we check for.

File: code/modules/mob/living/living.dm [hunk 4]
```diff
@@ -1167,3 +1167,36 @@
...
+		if(isturf(hit_atom))
+			var/turf/T = hit_atom
+			if(iswallturf(T))
+				T.dismantle_wall(TRUE)
+				hit_something = TRUE
+		if(hit_something)
+			visible_message("<span class='danger'>[src] slams into [hit_atom]!</span>", "<span class='userdanger'>You slam into [hit_atom]!</span>")
+			playsound(get_turf(src), 'sound/effects/meteorimpact.ogg', 100, TRUE)
+		return
+	if(has_status_effect(STATUS_EFFECT_IMPACT_IMMUNE))
```
Comment (lewcc) [inline code/modules/mob/living/living.dm:1201]: Move this up and remove the return.

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm [hunk 5]
```diff
@@ -40,3 +42,34 @@
...
+	else
+		RegisterSignal(src, COMSIG_MOB_LOGIN, TYPE_PROC_REF(/mob/living/simple_animal/hostile/poison/terror_spider/prince, give_spell))
+
+/mob/living/simple_animal/hostile/poison/terror_spider/prince/proc/give_spell()
+	SIGNAL_HANDLER
+	var/obj/effect/proc_holder/spell/spell = new /obj/effect/proc_holder/spell/princely_charge()
+	mind.AddSpell(spell)
+
+/obj/effect/proc_holder/spell/princely_charge
+	name = "Princely charge"
```
Comment (Henri215) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:60]:
```suggestion
	name = "Princely Charge"
```

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm [hunk 6]
```diff
@@ -40,3 +42,34 @@
...
 	L.adjustStaminaLoss(40)
 	return ..()
+
+/mob/living/simple_animal/hostile/poison/terror_spider/prince/Initialize(mapload)
+	. = ..()
+	if(mind)
+		var/obj/effect/proc_holder/spell/spell = new /obj/effect/proc_holder/spell/princely_charge()
+		mind.AddSpell(spell)
+	else
+		RegisterSignal(src, COMSIG_MOB_LOGIN, TYPE_PROC_REF(/mob/living/simple_animal/hostile/poison/terror_spider/prince, give_spell))
```
Comment (S34NW) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:52]: you might want to un-register this once the spell is given?

Comment (DGamerL) [reply code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:52]: I think I did it correctly like this?

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm [hunk 7]
```diff
@@ -40,3 +42,35 @@
...
```
Comment (Contrabang) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:52]: I believe this is the signal you're actually looking for, since we're basing this on minds and not clients
```suggestion
		RegisterSignal(src, COMSIG_BODY_TRANSFER_TO, TYPE_PROC_REF(/mob/living/simple_animal/hostile/poison/terror_spider/prince, give_spell))
```

Comment (DGamerL) [reply code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:52]: As discussed in discord, the current one works fine. I also tested aghosting in and out of a prince, and it did not duplicate the spell!

File: code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm [hunk 8]
```diff
@@ -40,3 +42,35 @@
...
+	mind.AddSpell(spell)
+	UnregisterSignal(src, COMSIG_MOB_LOGIN)
+
+/obj/effect/proc_holder/spell/princely_charge
+	name = "Princely Charge"
+	desc = "You charge at wherever you click on screen, dealing large amounts of damage, stunning and destroying walls and other objects."
+	gain_desc = "You can now charge at a target on screen, dealing massive damage and destroying structures."
+	base_cooldown = 30 SECONDS
+	clothes_req = FALSE
+	action_icon_state = "vampire_charge"
```
Comment (DGamerL) [inline code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:66]: This works for now, but should probably have it's own icon. If nobody has suggestions, I'll just make this into a prince icon

Comment (Sirryan2002) [reply code/modules/mob/living/simple_animal/hostile/terror_spiders/prince.dm:66]: I think prince icon is a fine stand in for production until someone gets around to making a spell icon for it

File: code/modules/mob/living/living.dm [hunk 9]
```diff
@@ -1167,3 +1167,35 @@
...
+	. = ..()
+	if(!has_status_effect(STATUS_EFFECT_CHARGING) || has_status_effect(STATUS_EFFECT_IMPACT_IMMUNE))
+		return
+
+	var/hit_something = FALSE
+	if(ismovable(hit_atom))
+		var/atom/movable/AM = hit_atom
+		var/atom/throw_target = get_edge_target_turf(AM, dir)
+		if(!AM.anchored || ismecha(AM))
+			AM.throw_at(throw_target, 5, 12, src)
```
Comment (Sirryan2002) [inline code/modules/mob/living/living.dm:1181]: is this supposed to be `!ismecha(AM)` ? You may wanna check move_force here instead

File: code/modules/mob/living/living.dm [hunk 10]
```diff
@@ -1167,3 +1167,35 @@
...
+		shake_camera(L, 4, 3)
+		hit_something = TRUE
+	if(isturf(hit_atom))
+		var/turf/T = hit_atom
+		if(iswallturf(T))
+			T.dismantle_wall(TRUE)
+			hit_something = TRUE
+	if(hit_something)
+		visible_message("<span class='danger'>[src] slams into [hit_atom]!</span>", "<span class='userdanger'>You slam into [hit_atom]!</span>")
+		playsound(get_turf(src), 'sound/effects/meteorimpact.ogg', 100, TRUE)
```
Comment (Sirryan2002) [inline code/modules/mob/living/living.dm:1201]: you may want to abstract this out into a prince_charge_impact proc. That way `throw_impact` proc for all living mobs isn't taken all up by a very specific use case.

File: code/modules/mob/living/living.dm (hunk 9 above)
Comment (GDNgit) [reply code/modules/mob/living/living.dm:1181]: mechas are actually anchored

Comment (DGamerL) [reply code/modules/mob/living/living.dm:1181]: I did copy this from the gargantua stuff, I think it is intended that mechs get thrown back?

File: code/modules/mob/living/living.dm (hunk 10 above)
Comment (DGamerL) [reply code/modules/mob/living/living.dm:1201]: Yeah this also makes me realize that this will proc will most likely be run twice by anything higher scoped than mob/living, will adjust!

**Review by Sirryan2002 — APPROVED** [review]
as long as this is well tested locally, this looks good to go. LGTM

**Review by S34NW — APPROVED** [review]
Deconflict for merge
